### PR TITLE
Fixed image refresh to avoid triggering a new refresh while the previ…

### LINKED ISSUE #420: Crash when chart refresh too low
I set the refresh to 10, no chart was shown and the app crashed after some time.
Version 1.8.0.8

### LINKED ISSUE #459: Image in sitemap
Image in sitemap not always shown and when it does it is slow and doesn't usually refresh

when accessing with chrome on android its fine

Image label="LivingCAM" url="http://172.16.10.106/zm/cgi-bin/nph-zms?mode=single&monitor=2&scale=50&user=xxxx&pass=xxx" refresh=500

### PR DESCRIPTION
**Pull request by lolodomo**
…ous is still in progress

A minimum refresh rate is set to 100 ms.

Fixes #420
Fixes #459

Signed-off-by: Laurent Garnier <lg.hc@free.fr>

### CODE REVIEW

**Comment by lolodomo** [comment]
Here are the logs showing what are the real refresh timestamps for a chart and a refresh set to 100 ms in my sitemap and with my server (RPI 2) requiring few seconds to generate the chart:
```
12-02 15:02:19.091 4803-4803/org.openhab.habdroid I/MySmartImageView: Refreshing image at https://192.168.1.36:8443/chart?groups=GHumiditeInt&period=W&random=-633635428&theme=black&dpi=320&w=2560&h=1280
12-02 15:02:21.842 4803-4803/org.openhab.habdroid I/MySmartImageView: Refreshing image at https://192.168.1.36:8443/chart?groups=GHumiditeInt&period=W&random=-633635428&theme=black&dpi=320&w=2560&h=1280
12-02 15:02:24.341 4803-4803/org.openhab.habdroid I/MySmartImageView: Refreshing image at https://192.168.1.36:8443/chart?groups=GHumiditeInt&period=W&random=-633635428&theme=black&dpi=320&w=2560&h=1280
12-02 15:02:26.842 4803-4803/org.openhab.habdroid I/MySmartImageView: Refreshing image at https://192.168.1.36:8443/chart?groups=GHumiditeInt&period=W&random=-633635428&theme=black&dpi=320&w=2560&h=1280
12-02 15:02:29.346 4803-4803/org.openhab.habdroid I/MySmartImageView: Refreshing image at https://192.168.1.36:8443/chart?groups=GHumiditeInt&period=W&random=-633635428&theme=black&dpi=320&w=2560&h=1280
12-02 15:02:31.591 4803-4803/org.openhab.habdroid I/MySmartImageView: Refreshing image at https://192.168.1.36:8443/chart?groups=GHumiditeInt&period=W&random=-633635428&theme=black&dpi=320&w=2560&h=1280
12-02 15:02:33.841 4803-4803/org.openhab.habdroid I/MySmartImageView: Refreshing image at https://192.168.1.36:8443/chart?groups=GHumiditeInt&period=W&random=-633635428&theme=black&dpi=320&w=2560&h=1280
12-02 15:02:36.342 4803-4803/org.openhab.habdroid I/MySmartImageView: Refreshing image at https://192.168.1.36:8443/chart?groups=GHumiditeInt&period=W&random=-633635428&theme=black&dpi=320&w=2560&h=1280
12-02 15:02:38.594 4803-4803/org.openhab.habdroid I/MySmartImageView: Refreshing image at https://192.168.1.36:8443/chart?groups=GHumiditeInt&period=W&random=-633635428&theme=black&dpi=320&w=2560&h=1280
12-02 15:02:40.851 4803-4803/org.openhab.habdroid I/MySmartImageView: Refreshing image at https://192.168.1.36:8443/chart?groups=GHumiditeInt&period=W&random=-633635428&theme=black&dpi=320&w=2560&h=1280
12-02 15:02:43.591 4803-4803/org.openhab.habdroid I/MySmartImageView: Refreshing image at https://192.168.1.36:8443/chart?groups=GHumiditeInt&period=W&random=-633635428&theme=black&dpi=320&w=2560&h=1280
12-02 15:02:46.105 4803-4803/org.openhab.habdroid I/MySmartImageView: Refreshing image at https://192.168.1.36:8443/chart?groups=GHumiditeInt&period=W&random=-633635428&theme=black&dpi=320&w=2560&h=1280
12-02 15:02:47.820 4803-4803/org.openhab.habdroid I/MySmartImageView: Refreshing image at https://192.168.1.36:8443/chart?groups=GHumiditeInt&period=W&random=1576941107&theme=black&dpi=320&w=2560&h=1280
12-02 15:02:50.070 4803-4803/org.openhab.habdroid I/MySmartImageView: Refreshing image at https://192.168.1.36:8443/chart?groups=GHumiditeInt&period=W&random=1576941107&theme=black&dpi=320&w=2560&h=1280
12-02 15:02:52.820 4803-4803/org.openhab.habdroid I/MySmartImageView: Refreshing image at https://192.168.1.36:8443/chart?groups=GHumiditeInt&period=W&random=1576941107&theme=black&dpi=320&w=2560&h=1280
12-02 15:02:55.070 4803-4803/org.openhab.habdroid I/MySmartImageView: Refreshing image at https://192.168.1.36:8443/chart?groups=GHumiditeInt&period=W&random=1576941107&theme=black&dpi=320&w=2560&h=1280
12-02 15:02:57.572 4803-4803/org.openhab.habdroid I/MySmartImageView: Refreshing image at https://192.168.1.36:8443/chart?groups=GHumiditeInt&period=W&random=1576941107&theme=black&dpi=320&w=2560&h=1280
12-02 15:03:00.069 4803-4803/org.openhab.habdroid I/MySmartImageView: Refreshing image at https://192.168.1.36:8443/chart?groups=GHumiditeInt&period=W&random=1576941107&theme=black&dpi=320&w=2560&h=1280
12-02 15:03:02.320 4803-4803/org.openhab.habdroid I/MySmartImageView: Refreshing image at https://192.168.1.36:8443/chart?groups=GHumiditeInt&period=W&random=1576941107&theme=black&dpi=320&w=2560&h=1280
12-02 15:03:04.820 4803-4803/org.openhab.habdroid I/MySmartImageView: Refreshing image at https://192.168.1.36:8443/chart?groups=GHumiditeInt&period=W&random=1576941107&theme=black&dpi=320&w=2560&h=1280
```
No crash.

**Comment by lolodomo** [comment]
You can see that the random value in the URL is changed at a certain time, it corresponds to a reload of the page (the now famous 30s reload).

**Comment by lolodomo** [comment]
@mueller-ma : as a side note, what do you think about this comment: https://github.com/openhab/openhab.android/blob/master/mobile/src/main/java/org/openhab/habdroid/ui/OpenHABWidgetAdapter.java#L511 ?
When the chart is refreshed, the drawable is not cleared. Is it a problem ?
At page initialisation, I see that the drawable is cleared in the case of a chart but not in the case of an image. I am curious to know why it is different and what is the correct one ?

**Comment by mueller-ma** [comment]
@lolodomo I once removed that line to see what happens. When you have a larger sitemap with multiple charts and you start scrolling, the chart images will get mixed.

File: mobile/src/main/java/org/openhab/habdroid/util/MySmartImageView.java [hunk 1]
```diff
@@ -36,64 +38,100 @@
...
     }
 
     public void setRefreshRate(int msec) {
-        Log.i(TAG, "Setting image refresh rate to " + msec + " msec for " + myImageUrl);
-        if (this.imageRefreshTimer != null) {
-            this.imageRefreshTimer.cancel();
+        cancelRefresh();
+
+        // Consider a minimal refresh rate of 100 ms
+        if (msec < 100) {
```
Comment (mueller-ma) [inline mobile/src/main/java/org/openhab/habdroid/util/MySmartImageView.java]: Maybe change that in model/OpenHABWidget setRefresh()

**Comment by lolodomo** [comment]
Review comment taken into account.

File: mobile/src/main/java/org/openhab/habdroid/model/OpenHABWidget.java [hunk 2]
```diff
@@ -174,8 +174,10 @@ public int getRefresh() {
 		return refresh;
 	}
 
-	public void setRefresh(int refresh) {
-		this.refresh = refresh;
+	public void setRefresh(int refresh)
+	{
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/model/OpenHABWidget.java]: I know this is a nitpick, but whitespace change seems unneeded here.

File: mobile/src/main/java/org/openhab/habdroid/model/OpenHABWidget.java [hunk 3]
```diff
@@ -174,8 +174,10 @@ public int getRefresh() {
...
+		// Consider a minimal refresh rate of 100 ms
+		this.refresh = (refresh < 100) ? 100 : refresh;
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/model/OpenHABWidget.java]: `this.refresh = Math.max(100, refresh);`

**Comment by lolodomo** [comment]
Code adjusted to consider @maniac103 's review comments.

**Pull request merged**